dask/operation.py

```diff
@@ -0,0 +1,384 @@
+from dask.core import reverse_dict
+from dask.highlevelgraph import HighLevelGraph
+from dask.optimization import SubgraphCallable
+
+
+@dataclass(frozen=True)
+class CollectionOperation:
+    """CollectionOperation class
+
+    Encapsulates the state and graph-creation logic for
+    a generic Dask collection.
+    """
+
+    @property
+    def name(self) -> str:
+        """Return the name for this CollectionOperation
+
+        This name should match the string label used in the
+        output keys of a task grah terminating with this operation.
+        """
+        raise NotImplementedError
+
+    @property
+    def dependencies(self) -> frozenset[CollectionOperation]:
+        """Return set of CollectionOperation dependencies"""
+        raise NotImplementedError
+
+    def subgraph(self, keys: list) -> tuple[dict, dict]:
+        """Return the subgraph and key dependencies for this operation
+
+        NOTE: Since this method returns a mapping between dependencies
+        and required keys, the logic may include recursively fusion.
+
+        Parameters
+        ----------
+        keys : list[KeyType]
+            List of required output keys needed from this collection
+
+        Returns
+        -------
+        graph : dict[tuple[Hashable], Any]
+            The subgraph for the current operation
+        dependencies : dict[CollectionOperation, list[Hashable]]
+            A dictionary mapping ``CollectionOperation`` objects
+            to required keys. This dictionary will be used by the
+            global graph-generation algorithm to determine which
+            operation-key combinations need to be materialized after
+            this operation.
+        """
+        raise NotImplementedError
+
+    def reinitialize(
+        self, replace_dependencies: dict[str, CollectionOperation], **changes
+    ) -> CollectionOperation:
+        """Reinitialize this CollectionOperation
+
+        Parameters
+        ----------
+        replace_dependencies : dict[str, CollectionOperation]
+            Replaced dependencies for the new operation
+        **changes : dict
+            New fields to use when initializing the new operation
+        """
+        raise NotImplementedError
+
+    @property
+    def collection_keys(self) -> list:
+        """Get the collection keys for this operation"""
+        raise NotImplementedError
+
+    def copy(self) -> CollectionOperation:
+        """Return a shallow copy of this operation"""
+        return copy.copy(self)
+
+    @cached_property
+    def dask(self) -> HighLevelGraph:
+        dsk, _ = self.subgraph(self.collection_keys)
+        return HighLevelGraph.from_collections(
+            self.name,
+            dsk,
+            dependencies=[],
+        )
+
+
+@dataclass(frozen=True)
+class LiteralInputs(CollectionOperation):
+    """LiteralInputs class
+
+    Defines literal block/partition inputs to a CollectionOperation
+    """
+
+    _inputs: Mapping
+    _label: str = "literals"
```

Comment:
Any thoughts about making these public? With immutability there doesn't seem to be much downside. I also think that introspecting `init()` arguments during replays is intended behavior.

Reply:
Sure - I suppose we only need the underscore for field names that clash with required properties.